linenoise.hpp

```diff
@@ -1940,151 +2031,166 @@ inline void refreshMultiLine(struct linenoiseState *l) {
-                refreshLine(l);
+                RefreshLine();
             }
         } else {
-            memmove(l->buf+l->pos+clen,l->buf+l->pos,l->len-l->pos);
-            memcpy(&l->buf[l->pos],cbuf,clen);
-            l->pos+=clen;
-            l->len+=clen;
-            l->buf[l->len] = '\0';
-            refreshLine(l);
+            memmove(buf+pos+clen,buf+pos,len-pos);
+            memcpy(&buf[pos],cbuf,clen);
+            pos+=clen;
+            len+=clen;
+            buf[len] = '\0';
+            RefreshLine();
         }
     }
     return 0;
 }
 
 /* Move cursor on the left. */
-inline void linenoiseEditMoveLeft(struct linenoiseState *l) {
-    if (l->pos > 0) {
-        l->pos -= unicodePrevGraphemeLen(l->buf, l->pos);
-        refreshLine(l);
+void linenoiseState::linenoiseEditMoveLeft() {
+    if (pos > 0) {
+        pos -= unicodePrevGraphemeLen(buf, pos);
+        RefreshLine();
     }
 }
 
 /* Move cursor on the right. */
-inline void linenoiseEditMoveRight(struct linenoiseState *l) {
-    if (l->pos != l->len) {
-        l->pos += unicodeGraphemeLen(l->buf, l->len, l->pos);
-        refreshLine(l);
+void linenoiseState::linenoiseEditMoveRight() {
+    if (pos != len) {
+        pos += unicodeGraphemeLen(buf, len, pos);
+        RefreshLine();
     }
 }
 
 /* Move cursor to the start of the line. */
-inline void linenoiseEditMoveHome(struct linenoiseState *l) {
-    if (l->pos != 0) {
-        l->pos = 0;
-        refreshLine(l);
+void linenoiseState::linenoiseEditMoveHome() {
+    if (pos != 0) {
+        pos = 0;
+        RefreshLine();
     }
 }
 
 /* Move cursor to the end of the line. */
-inline void linenoiseEditMoveEnd(struct linenoiseState *l) {
-    if (l->pos != l->len) {
-        l->pos = l->len;
-        refreshLine(l);
+void linenoiseState::linenoiseEditMoveEnd() {
+    if (pos != len) {
+        pos = len;
+        RefreshLine();
     }
 }
 
 /* Substitute the currently edited line with the next or previous history
  * entry as specified by 'dir'. */
 #define LINENOISE_HISTORY_NEXT 0
 #define LINENOISE_HISTORY_PREV 1
-inline void linenoiseEditHistoryNext(struct linenoiseState *l, int dir) {
-    if (history.size() > 1) {
-        /* Update the current history entry before to
-         * overwrite it with the next one. */
-        history[history.size() - 1 - l->history_index] = l->buf;
-        /* Show the new entry */
-        l->history_index += (dir == LINENOISE_HISTORY_PREV) ? 1 : -1;
-        if (l->history_index < 0) {
-            l->history_index = 0;
-            return;
-        } else if (l->history_index >= (int)history.size()) {
-            l->history_index = static_cast<int>(history.size())-1;
-            return;
-        }
-        memset(l->buf, 0, l->buflen);
-        strcpy(l->buf,history[history.size() - 1 - l->history_index].c_str());
-        l->len = l->pos = static_cast<int>(strlen(l->buf));
-        refreshLine(l);
+void linenoiseState::linenoiseEditHistoryNext(int dir) {
```

Comment:
The new history handling code looks very different from the original linenoise code of antires. Does it have the same behavior? If not, how does it work?

Reply:
It's similar but probably not quite identical.  I was trying to make it easier to reason about - the "working" buffer is stashed in a temporary buffer rather than the history vector, since its contents (unlike the rest of the history) may change up until the line is entered.  Also, we avoid the need for the initial "AddHistory("")" which felt a little strange to me - in the current approach the indexing and data reflect the fact that there *is* no history until something is entered.

Reply:
If the new behavior is different from the original, could you please put it in another pull request? This way, we can ensure that this pull request doesn't cause any behavior changes. I would like to handle things one at a time. Thanks for your understanding!

Reply:
OK, I've backed them out.